PR1/5 - refactor(bookmark) : create bookmark documents and repo (#877)

itsAidenDev opened the pull request:
This PR corresponds to [Task 877](https://tree.taiga.io/project/luisvi-ita-challenges/task/877) and introduces the first step of a broader refactor aimed at decoupling bookmark logic from the User and Challenge microservices. The goal of this step is to establish the domain, persistence layer, and base structures for the future bookmark service.

### 🔹 Key Changes:

- **Extracted common fields into a new InteractionDocument, now extended by both BookmarkDocument and FavoriteDocument.**
- Added the Bookmark domain and persistence layer.
- Created BookmarkDocument under com.itachallenge.userinteraction.document.bookmark
- Implemented BookmarkRepository using ReactiveMongoRepository.
- Added BookmarkResponseDto.


The new bookmarks collection is initialized as empty. For now, the GET endpoint continues to rely on the Bookmark field within UserDocument.

These changes do not modify any existing API endpoints or response formats.

The BookmarkResponseDto remains unused at this stage (and will also be unused in the related PRs), since the front-end currently expects the GET endpoint to return an array of UUID.


### This task is the first of a serie of 5:

[#877](https://github.com/IT-Academy-BCN/ita-challenges-backend/compare/refactor/project/luisvi-ita-challenges/t/877)  Create "Bookmark" documents and repositories (new collection) : **_THIS PR_**

[#878](https://github.com/IT-Academy-BCN/ita-challenges-backend/compare/refactor/project/luisvi-ita-challenges/t/878)  Implement "BookmarkService" (the logic)

[#879](https://github.com/IT-Academy-BCN/ita-challenges-backend/compare/refactor/project/luisvi-ita-challenges/t/879) Implement "BookmarkController" (getting user's bookmarks from new collection)

[#880](https://github.com/IT-Academy-BCN/ita-challenges-backend/compare/refactor/project/luisvi-ita-challenges/t/880)  Integration tests + Documentation update

[#881](https://github.com/IT-Academy-BCN/ita-challenges-backend/compare/refactor/project/luisvi-ita-challenges/t/881) Delete field "bookmarkChallenges" from the UserDocument


### ✅ Created files

**Documents & DTOs:**
- InteractionDocument
- BookmarkDocument & BookmarkDocumentTest
- BookmarkResponseDto & BookmarkResponseDtoTest

**Repository:**
- BookmarkRepository

### ✅ Modified files
- FavoriteDocument
- changelog.md


### Changelog updated

### PR Author Self-Check

- [x]  New BookmarkDocument entity created under the bookmarks collection.
- [x]  BookmarkRepository created.
- [x]  DTO created
- [x]  Changelog updated
- [x]  Code follows agreed-upon project conventions (naming, formatting, Sonar rules).
- [x]  All tests pass locally and in the development environment.
- [x]  SonarQube/SonarCloud analysis passes with no new issues or code smells.

itsAidenDev committed:
Created bookmark docs and repo

itsAidenDev committed:
Created bookmark docs and repo

itsAidenDev committed:
Added more testing to BookmarkResponseDtoTest

itsAidenDev committed:
Added more testing to BookmarkResponseDtoTest

itsAidenDev committed:
Reversed changes in UserServiceImpl

itsAidenDev committed:
Reversed changes in UserServiceImpl

zohra-b commented:
This PR is consistent with what have been done is the ultimate sprint with "Favorites". Approved

DiegoBalaguer reviewed (approved):
* BookmarkResponseDto
   Estais usando Data y Getter de Lombok. Realmente si estais usando Data ya lleva implicito el Getter.

* BockmarkDocument
   En los nombres para los campos de la Collection de Mongo estais usando formato CamelCase cuando en realidad tendria que ser snake_case.

FlavioKde reviewed (approved):
Jose Luis and Zohra, good work. I'm leaving some adjustment comments, approved.

BookmarkDocumentTest.java is missing the following in this method: @Test

void equals_WithNullUserIdInBoth_test()

In BookmarkResponseDto, there's redundant code:

@Data
@Builder
@Getter

In BookmarkDocument, for a complete audit:

@CreatedDate
@Field(name = "createdAt")
private LocalDateTime createdAt;

You must add:
@LastModifiedDate
@Field(name = "updatedAt")
private LocalDateTime updatedAt

Good work

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/dto/userinteraction/bookmark/BookmarkResponseDto.java:
```diff
@@ -0,0 +1,27 @@
+@NoArgsConstructor
+public class BookmarkResponseDto {
+
+    @JsonProperty(value = "uuid_favorite")
```
uuid_bookmark?

zohra-b replied:
done

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/dto/userinteraction/bookmark/BookmarkResponseDtoTest.java:
```diff
@@ -0,0 +1,267 @@
+        try {
+            java.lang.reflect.Field uuidField = BookmarkResponseDto.class.getDeclaredField("uuid");
+            com.fasterxml.jackson.annotation.JsonProperty uuidAnnotation = uuidField.getAnnotation(com.fasterxml.jackson.annotation.JsonProperty.class);
+            assertEquals("uuid_favorite", uuidAnnotation.value());
```
uuid_bookmark?

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/dto/userinteraction/bookmark/BookmarkResponseDtoTest.java:
The testing effort is disproportionate for a simple DTO with Lombok. We don't need to test all the details of equals/hashCode/toString/builder, because that's Lombok's responsibility.

The important thing here is:

that we can construct the DTO with the correct values, and

that the JSON properties (uuid_bookmark, user_id, challenge_id, created_at) are those defined in the contract (ideally validated with actual serialization/deserialization using ObjectMapper).

I suggest simplifying the testing: keep 1–2 tests focused on this and remove the rest of the redundant Lombok tests.

zohra-b replied:
Sonar won't accept 1 or 2 tests. Or we should quit @data annotation. That'a was has benn validated for favorites

ivilarop replied:
Sonar doesn't impose a minimum number of tests, but rather enforces code coverage.

We don't need more than 20 tests that reimplement Lombok's behavior; we only need a few tests that meaningfully exercise the DTO.

Removing `@Data` would be a step backward: it adds repetitive text and doesn't improve quality. Lombok isn't the problem.

For this DTO, something like this is sufficient:

One test that builds the DTO (builder / all-args) and checks the field values.

One or two tests that serialize/deserialize with ObjectMapper and validate the JSON properties (uuid_bookmark, user_id, challenge_id, created_at).

These tests:

provide real code coverage for Sonar,

validate the API contract,

and avoid a large number of low-value Lombok tests.

What was done for `favorite` works, but it's excessive, and approving it was a mistake. Here we are trying to move towards a more realistic testing strategy: focusing on behavior and contracts, not retesting what Lombok already guarantees.

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/userinteraction/document/bookmark/BookmarkDocumentTest.java:
The BookmarkDocument tests are very detailed for what is essentially a simple @Document with Lombok.

Just like with the DTO, we don't need to thoroughly test equals/hashCode/toString/builder, because Lombok generates these and they aren't part of the domain logic. Maintaining so many tests for each POJO makes the project difficult to maintain without adding real value.

In this case, something much more minimalist would suffice:

A test that verifies the document is constructed correctly (builder or constructor with arguments) and that the getters return the expected values.

Optionally, a test of the no-args constructor if we want to ensure the framework can instantiate it.

If we want to add more advanced tests at some point, it would make more sense to perform repository integration tests (saving/reading to MongoDB and verifying proper persistence) than to test every combination of equals/hashCode/toString.

zohra-b replied:
An InteractionDocument has been created, extended by both BookmarkDocument and FavoriteDocument, in order to avoid Sonarcloud issue about code duplicated

Also tests have been reduced and we implemented one songle test for hash / equals / toString methods

zohra-b committed:
Change JSON property name from 'uuid_favorite' to 'uuid_bookmark'

zohra-b committed:
Change JSON property name from 'uuid_favorite' to 'uuid_bookmark'

zohra-b committed:
Update JsonProperty value for uuid field in tests

zohra-b committed:
Update JsonProperty value for uuid field in tests

zohra-b committed:
Refactor BookmarkDocumentTest by removing redundant tests

Removed multiple tests related to equals, hashCode, and builder methods for BookmarkDocument.

zohra-b committed:
Refactor BookmarkDocumentTest by removing redundant tests

Removed multiple tests related to equals, hashCode, and builder methods for BookmarkDocument.

zohra-b committed:
Remove unused bookmarkDocument2 in tests

Removed unused bookmarkDocument2 initialization from tests.

zohra-b committed:
Remove unused bookmarkDocument2 in tests

Removed unused bookmarkDocument2 initialization from tests.

zohra-b committed:
Remove redundant equality and hashCode tests

Removed multiple tests related to equals, hashCode, and builder methods from BookmarkResponseDtoTest.

zohra-b committed:
Remove redundant equality and hashCode tests

Removed multiple tests related to equals, hashCode, and builder methods from BookmarkResponseDtoTest.

zohra-b committed:
Quitted @data to avoid Sonar request coverage

zohra-b committed:
Quitted @data to avoid Sonar request coverage

zohra-b committed:
Add missing closing brace in BookmarkDocumentTest

zohra-b committed:
Add missing closing brace in BookmarkDocumentTest

aperez1haad committed:
Revert "Replace data for getter/setter"

This reverts commit fb0a57fb77879f431e9e22e974d63005c40f4ff3.

aperez1haad committed:
Revert "Replace data for getter/setter"

This reverts commit fb0a57fb77879f431e9e22e974d63005c40f4ff3.

eduzavarce committed:
Refactor: Extract common fields to InteractionDocument and extend in BookmarkDocument and FavoriteDocument

eduzavarce committed:
Refactor: Extract common fields to InteractionDocument and extend in BookmarkDocument and FavoriteDocument

eduzavarce committed:
Remove unused @AllArgsConstructor import from BookmarkDocument

eduzavarce committed:
Remove unused @AllArgsConstructor import from BookmarkDocument

eduzavarce committed:
Remove redundant assertions and add missing test annotation in FavoriteDocumentTest

eduzavarce committed:
Remove redundant assertions and add missing test annotation in FavoriteDocumentTest

eduzavarce committed:
Consolidate and refactor equals, hashCode, and toString tests in BookmarkDocumentTest and FavoriteDocumentTest

eduzavarce committed:
Consolidate and refactor equals, hashCode, and toString tests in BookmarkDocumentTest and FavoriteDocumentTest

zohra-b committed:
updated changelog.md

zohra-b committed:
updated changelog.md

zohra-b committed:
Fix typo in CHANGELOG for InteractionDocument

zohra-b committed:
rebase with develop

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/userinteraction/document/InteractionDocument.java:
1. Mark the class as abstract.
2. Since it's an abstract base class and uses `@SuperBuilder`, `@AllArgsConstructor` is probably unnecessary. It's better to limit public constructors to prevent direct instantiation.

3. Do we really need `protected`, or can we keep the fields private and use Lombok's getters/setters? This way, we maintain better encapsulation.

4. In `@EqualsAndHashCode`, it might be better to include only the UUID as the document's identifier to avoid problems if `userId` or `challengeId` change.

5. Optional: We could group it into a `document.interaction` subpackage to make it clear that it's the foundation of this type of document.

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/dto/userinteraction/bookmark/BookmarkResponseDtoTest.java:
The tests are correct and cover many cases, but for a DTO with Lombok, you've gone a bit overboard 😄.

– There's no need to test all the constructors/getters/setters: that's Lombok-generated code and adds little value.

– What does make sense to test is the JSON contract (JsonProperty), ideally with a real serialization test using ObjectMapper.

We could stick with:

1 test that validates the builder (creates the DTO and checks the fields).

1 test that serializes to JSON and verifies that the keys are uuid_bookmark, user_id, etc.

With that, you have good coverage and tests more focused on business value than implementation details.

ivilarop replied:
done

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/userinteraction/document/bookmark/BookmarkDocumentTest.java:
We generally don't test Lombok (constructors, getters, setters) because these tests are low-value. The important parts of the documentation will be logic (validations, additional methods, repository behavior, etc.). Here, 1-2 tests focused on the builder and equals/hashcode are sufficient.

zohra-b replied:
I simplified as much as I could : 2 tests remaining. Quitting the getters and setters test makes the coverage fall to 50%, so I had to keep it

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/userinteraction/document/favorite/FavoriteDocumentTest.java:
I understand you've added many constructor/getter/setter tests to increase Sonar's coverage. Technically, it works, but in real-world projects, code generated by Lombok isn't usually tested because it adds little value.

What we do need is:
- a test that creates the BookmarkDocument/FavoriteDocument using the builder and checks the fields,
- optionally, an equals/hashCode test if identity is important.

With that, we already have reasonable coverage of these documents. We should gain the remaining coverage by testing service/repository logic, not all the getters and constructors.

We'll review it to see what the problem is.

zohra-b replied:
FavoriteDocumentTests is part of the previous UserStory. It is not in the scope of this US.

zohra-b replied:
As soon as you'll approve the Booksmark changes, I will transpose them to Favorites  service

zohra-b committed:
made InteractionDocument class abstract

zohra-b committed:
quitted AllArgConstructor annotation

zohra-b committed:
turned InteractionDocument fields private (not protected) and turned to add AllArgConstructors for build was unsuccessful without

zohra-b committed:
quitted constructors in favoritedocument and in Bookmarkdocument to use SuperBuilder

zohra-b committed:
modified entire BookmarkDocumentTest class to maintain one only test

zohra-b committed:
quitted no arg / all arg test - coverage still 100

zohra-b committed:
simplified BookmarkResponseDto test : 2 tests remainings

zohra-b committed:
refactor BookmarkdocumentTest

zohra-b committed:
refactored FavoriteDocumentTest : one only test remaining

zohra-b committed:
refactored FavoriteResponseDtoTest : one test, coverage 100%

ivilarop committed:
test(userinteraction): simplify document tests and keep coverage

ivilarop committed:
test(userinteraction): simplify document tests and keep coverage

ivilarop committed:
test(userinteraction): simplify document tests and keep coverage
